corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,329 @@
-class UserChangeMessage(object):
+from enum import Enum
+from django.utils.translation import ugettext as _
+
+
+class Change(Enum):
+    SET = 'set'
+    RESET = 'reset'
+    REMOVE = 'remove'
+    ADD = 'add'
+
+
+class UserChangeMessageV1(object):
```

Comment:
A new version would make sense if there is a structural change in the message that needs different/additional information saved in the log saved for it or if the log is modified to add more information that we want to show to the user now.

I don't think `UserChangeMessageV1` would/should be changed at all once set, only formatter could be tweaked like if we were do a change that is only related to what message we show,  like using "project" instead of "domain", which does not need a new formatter version.

For example, if for domain change we want to also keep its subscription detail/name at the time of the change. Or include permissions available with the role at the time of assignment.

I imagine this would be like 
```
class UserChangeMessageV1(object):
    version = 1  # None as of now

    @staticmethod
    def domain_change(domain):
        return {"domain": {"action": {"name": domain}}}
    
    @staticmethod
    def role_change(role):
        return {"role": {"action": {"name": role.name}}}
    
    @staticmethod
    def other_change():
        return {"foo": "bar"}
    # ... other methods


class UserChangeMessageV2(UserChangeMessageV1):
    version = 2
    
    @staticmethod
    def domain_change(domain, subscription):
        return {"domain": {"action": {"name": domain, "subscription": "Pro"}}}
    
    @classmethod
    def role_change(cls, role, permissions):
        change_message = super().role_change(role)
        # modify change message to include permissions
        return change_message
```

>  In my opinion it makes more sense to version the individual messages themselves (the class methods) as they're what's likely to change.

```
class UserChangeMessage(object):
    def domain_change(domain):
        # pass

    def domain_change_v2(domain, permissions):
        # pass
```
Are you suggesting like this?
How will one determine the version to be read? Happy to hear more on this.

Also,
I like how classes would be shorter and have a clear versioning. This class here would just keep getting longer right?

Reply:
Once there's a new version of `domain_change`, do we have any reason to keep the old one around?  We wouldn't use it anymore, would we?  We'd just need to keep the old version of the formatter around to be able to render the messages already in the DB, and do something to bump the version number so we know which formatter version to use.

Reply:
Yes, You are right, we would just need the formmatter to display old messages. We could simply remove V1 if we write a fresh V2 (without inheriting from V1).

Reply:
I don't see why we'd need to do any kind of inheritance at all. We'd just add a v2 handler for the specific message that changed.